src/frontend/src/webhook/mod.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::net::{IpAddr, SocketAddr};
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context};
+use axum::body::Bytes;
+use axum::extract::{Extension, Path};
+use axum::http::{HeaderMap, Method, StatusCode};
+use axum::routing::post;
+use axum::Router;
+use pgwire::net::Address;
+use pgwire::pg_server::SessionManager;
+use risingwave_common::secret::LocalSecretManager;
+use risingwave_sqlparser::ast::{Expr, ObjectName};
+use tokio::net::TcpListener;
+use tower::ServiceBuilder;
+use tower_http::add_extension::AddExtensionLayer;
+use tower_http::compression::CompressionLayer;
+use tower_http::cors::{self, CorsLayer};
+
+use crate::handler::handle;
+use crate::webhook::utils::{err, Result};
+mod utils;
+
+pub type Service = Arc<WebhookService>;
+
+// We always use the `root` user to connect to the database to allow the webhook service to access all tables.
+const USER: &str = "root";
+
+#[derive(Clone)]
+pub struct WebhookService {
+    webhook_addr: SocketAddr,
+}
+
+pub(super) mod handlers {
+    use std::net::Ipv4Addr;
+
+    use risingwave_pb::catalog::WebhookSourceInfo;
+    use risingwave_sqlparser::ast::{Query, SetExpr, Statement, Value, Values};
+    use utils::{header_map_to_json, verify_signature};
+
+    use super::*;
+    use crate::catalog::root_catalog::SchemaPath;
+    use crate::session::SESSION_MANAGER;
+
+    pub async fn handle_post_request(
+        Extension(_srv): Extension<Service>,
+        headers: HeaderMap,
+        Path((database, schema, table)): Path<(String, String, String)>,
+        body: Bytes,
+    ) -> Result<()> {
+        let session_mgr = SESSION_MANAGER
+            .get()
+            .expect("session manager has been initialized");
+
+        // Can be any address, we use the port of meta to indicate that it's a internal request.
+        let dummy_addr = Address::Tcp(SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 5691));
+
+        // TODO(kexiang): optimize this
+        // get a session object for the corresponding database
+        let session = session_mgr
```

Comment:
Shall we cache the session object & built verification expression to avoid constructing them per record?

Reply:
> // TODO(kexiang): optimize this

Yes. Let's leave it as the following PR.